### PR TITLE
fix(shared): prevent double counting in realized expenses

### LINKED ISSUE #291: Corriger le double comptage des dépenses réalisées (enveloppes)
### Contexte

Un early adopter a remonté un bug : quand une prévision (budget line) ET ses transactions sont cochées, le montant réalisé est doublé. Exemple : une prévision de 50 CHF avec 2 transactions de 25 CHF cochées → le système affiche 100 CHF au lieu de 50 CHF. La fonction `calculateRealizedExpenses` additionne naïvement les prévisions cochées + les transactions cochées sans appliquer la logique d'enveloppe.

### Critères d'acceptation

- [x] CA1: Une prévision cochée de 50 CHF avec 2 transactions cochées de 25 CHF = 50 CHF de dépenses réalisées (pas 100 CHF)
- [x] CA2: Une prévision cochée sans transactions = montant de la prévision comptée comme réalisé
- [x] CA3: Les transactions libres (sans `budgetLineId`) cochées sont comptées directement dans les dépenses réalisées
- [x] CA4: Le nombre de transactions affiché ne double-compte pas les transactions allouées à une enveloppe cochée
- [x] CA5: Les tests unitaires couvrent les cas enveloppe dans `calculateRealizedExpenses`
- [x] CA6: Les transactions allouées cochées sont comptabilisées même si la prévision parente n'est pas cochée. Quand le parent est coché, `max(enveloppe, consommé)` s'applique

### Règles métier

- La logique d'enveloppe s'applique : pour une prévision cochée, on utilise `max(montant_enveloppe, montant_consommé_par_transactions)` — cohérent avec RG-005
- Les transactions allouées à une prévision cochée ne doivent pas être comptées une deuxième fois
- Les transactions libres (non allouées) restent comptées directement

### Notes techniques

- **Package(s) concerné(s)** : `shared`, `frontend`, `ios`
- La logique correcte existe déjà dans `calculateTotalExpensesWithEnvelopes` (`shared/src/calculators/budget-formulas.ts:114-144`) — s'en inspirer
- La fonction buggée est `calculateRealizedExpenses` (`shared/src/calculators/budget-formulas.ts:179-199`)
- Le signal `realizedExpenses` dans `budget-details-store.ts:225-232` consomme cette fonction
- Ajouter des tests unitaires dans `budget-formulas.spec.ts` pour les cas enveloppe réalisée

### Hors périmètre

- Refactoring global des formules de calcul
- Modification de la logique de cochage (check/uncheck)
- Impact sur le calcul du rollover

---

**Estimation : S**

### PR DESCRIPTION
**Pull request by neogenz**
## Summary
Apply envelope-aware accounting to `calculateRealizedExpenses`: checked budget lines now use `max(envelope_amount, consumed_by_transactions)` to prevent double counting. Allocated transactions are counted only via their envelope; free transactions are counted directly. Resolves the bug where a 50 CHF envelope with 2×25 CHF transactions was counted as 100 CHF.

## What Changed
- Refactored `calculateRealizedExpenses` to apply the same envelope logic as `calculateTotalExpensesWithEnvelopes`
- Updated type signatures to require `FinancialItemWithId` and `TransactionWithBudgetLineId`
- Added 11 new tests covering all acceptance criteria and edge cases (AC1–AC5)

## Testing
All 115 unit tests pass. New tests validate envelope logic, free transactions, rollover exclusion, and savings treatment.

### CODE REVIEW

File: backend-nest/src/modules/budget-line/budget-line.service.ts [hunk 1]
```diff
@@ -546,6 +549,27 @@ export class BudgetLineService {
         user,
       );
 
+      // Cascade: toggle all allocated transactions in the same direction
+      const { error: cascadeError } = await supabase
+        .from('transaction')
+        .update({ checked_at: newCheckedAt, updated_at: now })
+        .eq('budget_line_id', id);
```
[the remark this answers is not in the document]
Comment (neogenz) [reply backend-nest/src/modules/budget-line/budget-line.service.ts]: Ce commentaire est incorrect. La cascade ne peut pas échouer indépendamment du toggle car **tout se passe dans une seule fonction SQL `toggle_budget_line_check`**. En PostgreSQL, chaque appel à une fonction `SECURITY DEFINER` s'exécute dans une transaction implicite. Si l'UPDATE des transactions (lignes 42-45) échoue, l'UPDATE du budget line (lignes 36-39) est aussi rollback. Il n'y a aucun risque d'état incohérent.

File: backend-nest/supabase/migrations/20260201165036_add_toggle_budget_line_check_rpc.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    v_new_checked_at := v_now;
+  END IF;
+
+  -- Update budget line
+  UPDATE public.budget_line
+  SET checked_at = v_new_checked_at,
+      updated_at = v_now
+  WHERE id = p_budget_line_id;
+
+  -- Cascade to all allocated transactions
```
[the remark this answers is not in the document]
Comment (neogenz) [reply backend-nest/supabase/migrations/20260201165036_add_toggle_budget_line_check_rpc.sql:41]: Ce commentaire est incorrect. En PostgreSQL, **chaque fonction PL/pgSQL s'exécute automatiquement dans une transaction implicite**. Un `BEGIN...COMMIT` explicite n'est ni nécessaire ni même possible dans une fonction `SECURITY DEFINER`. Si n'importe quel statement échoue, toute la fonction est rollback automatiquement. C'est le comportement standard de PostgreSQL.

File: backend-nest/supabase/migrations/20260201165036_add_toggle_budget_line_check_rpc.sql [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+BEGIN
+  -- Get current checked_at to determine toggle direction
+  -- Ownership verified via monthly_budget.user_id (same as RLS policy)
+  -- FOR UPDATE locks the row to prevent concurrent toggle race conditions
+  SELECT bl.checked_at INTO v_current_checked_at
+  FROM public.budget_line bl
+  JOIN public.monthly_budget mb ON mb.id = bl.budget_id
+  WHERE bl.id = p_budget_line_id
+    AND mb.user_id = auth.uid()
+  FOR UPDATE OF bl;
```
[the remark this answers is not in the document]
Comment (neogenz) [reply backend-nest/supabase/migrations/20260201165036_add_toggle_budget_line_check_rpc.sql:22]: Ce risque n'existe pas. Chaque transaction a un seul `budget_line_id` — une transaction ne peut pas appartenir à deux budget lines simultanément. Deux utilisateurs qui togglent deux budget lines différents modifient donc des ensembles de transactions disjoints. Le `FOR UPDATE OF bl` suffit pour prévenir les race conditions sur le toggle d'une même budget line.

File: shared/src/calculators/budget-formulas.ts [hunk 4]
```diff
@@ -167,35 +167,59 @@ export class BudgetFormulas {
...
-      )
-      .reduce((sum, t) => sum + t.amount, 0);
+    // Pour chaque prévision cochée de type expense/saving, utiliser max(enveloppe, consommé)
+    budgetLines.forEach((line) => {
+      if (
+        line.checkedAt != null &&
+        (line.kind === 'expense' || line.kind === 'saving')
+      ) {
+        // Ignorer les lignes virtuelles de rollover
+        if (line.id.startsWith('rollover-')) return;
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts]: Over-engineering. Ajouter un flag `isVirtual` et un type guard pour un simple `startsWith('rollover-')` utilisé à un seul endroit viole KISS/YAGNI. Les lignes rollover sont un concept interne au frontend, pas un type métier. Si la convention change, un simple find-and-replace suffit.

[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts]: Même réponse que le commentaire précédent — c'est de l'over-engineering. `startsWith('rollover-')` est clair, utilisé à un seul endroit, et ne justifie pas l'ajout d'un type ou d'un flag dédié.

File: frontend/projects/webapp/src/app/feature/budget/budget-details/budget-details-page.ts [hunk 5]
```diff
@@ -302,7 +302,38 @@ export default class BudgetDetailsPage {
   }
 
   async handleToggleCheck(budgetLineId: string): Promise<void> {
+    if (budgetLineId.startsWith('rollover')) {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply frontend/projects/webapp/src/app/feature/budget/budget-details/budget-details-page.ts:310]: Over-engineering. Extraire `'rollover'` dans une constante `ROLLOVER_ID_PREFIX` pour un seul usage n'apporte pas de valeur. Le string est self-documenting et un find-and-replace couvre le risque de changement. Respectons KISS/YAGNI.

File: backend-nest/supabase/migrations/20260209120000_remove_transaction_cascade_from_toggle_check.sql [hunk 6]
```diff
@@ -0,0 +1,51 @@
+-- Remove transaction cascade from toggle_budget_line_check RPC.
+-- Transaction check state is now managed by the application layer (optimistic update)
+-- to allow independent checking of envelopes and their transactions.
```
[the remark this answers is not in the document]
Comment (neogenz) [reply backend-nest/supabase/migrations/20260209120000_remove_transaction_cascade_from_toggle_check.sql]: Bonne remarque, c'est fait. Le commentaire SQL a été enrichi avec le contexte métier :

```sql
-- Context: The envelope logic requires independent checking of budget lines and transactions.
-- Previously, checking a budget line would cascade to all its transactions, preventing users
-- from tracking partial expense realization (e.g., 50 CHF envelope with only 25 CHF spent).
-- Transaction check state is now managed by the application layer (optimistic update in frontend store).
```

Commit à venir.

File: backend-nest/src/modules/budget-line/budget-line.service.spec.ts [hunk 7]
```diff
@@ -540,24 +542,11 @@ describe('BudgetLineService', () => {
...
-          data: {
-            ...mockBudgetLineDb,
-            checked_at: checkedTimestamp,
-          },
+      mockSupabase.rpc.mockReturnValue({
+        single: jest.fn().mockResolvedValue({
+          data: { ...mockBudgetLineDb, checked_at: checkedTimestamp },
           error: null,
-        });
+        }),
```
[the remark this answers is not in the document]
Comment (neogenz) [reply backend-nest/src/modules/budget-line/budget-line.service.spec.ts:640]: Assertion RPC ajoutée dans les deux cas de test (check et uncheck) :

```typescript
expect(mockSupabase.rpc).toHaveBeenCalledWith(
  'toggle_budget_line_check',
  { p_budget_line_id: budgetLineId },
);
```

Le cas d'erreur RPC (ligne 582-595) était déjà couvert. Tests verts ✅

File: shared/src/calculators/budget-formulas.ts [hunk 8]
```diff
@@ -40,6 +40,7 @@ interface FinancialItem {
  */
 interface FinancialItemWithId extends FinancialItem {
   id: string;
+  isRollover?: boolean;
 }
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:44]: On décline cette suggestion :

- `isRollover?: boolean` optionnel est intentionnel. Cette propriété n'existe que dans le contexte des lignes rollover (ajoutées dynamiquement côté frontend). Forcer `false` par défaut sur toutes les lignes ajouterait du bruit sans valeur.

- `budgetLineId?: string | null` reflète la réalité Supabase/PostgreSQL : `null` = colonne existante mais vide en DB, `undefined` = propriété absente dans le mapping JS. Les deux états ont une sémantique différente et c'est un pattern standard avec Supabase.

Le typage actuel est correct et cohérent avec l'usage.

File: shared/src/calculators/budget-formulas.ts [hunk 9]
```diff
@@ -207,8 +228,8 @@ export class BudgetFormulas {
    * @returns Solde calculé depuis les éléments cochés uniquement
    */
   static calculateRealizedBalance(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:230]: Risque de drift noté. Un test de conformité cross-platform (TS vs Swift) n'est pas réaliste dans notre stack (pas de runtime partagé).

Mitigation actuelle :
1. Les deux fichiers ont des commentaires de sync renvoyant l'un vers l'autre
2. Les tests unitaires des deux côtés utilisent des scénarios équivalents (même inputs/outputs)

On documente cette contrainte mais on ne bloque pas la PR pour ça.

File: frontend/projects/webapp/src/app/feature/budget/budget-details/store/budget-details-check.utils.spec.ts [hunk 10]
```diff
@@ -199,18 +93,30 @@ describe('Budget Details Check Utils', () => {
       expect(result).not.toBeNull();
       expect(result!.isChecking).toBe(false);
       expect(result!.updatedBudgetLines[0].checkedAt).toBeNull();
-      expect(
-        result!.updatedTransactions.every((tx) => tx.checkedAt === null),
-      ).toBe(true);
-      expect(result!.transactionsToToggle).toHaveLength(2);
+      expect(result!.updatedTransactions[0].checkedAt).toBe(checkedAt);
     });
```
[the remark this answers is not in the document]
Comment (neogenz) [reply frontend/projects/webapp/src/app/feature/budget/budget-details/store/budget-details-check.utils.spec.ts:97]: Le rollback optimistic est déjà géré dans le store (`budget-details-store.ts` lignes 583-591) : en cas d'erreur API, l'état est recalculé depuis la réponse serveur via `patchState`. Le pattern est couvert par les tests e2e. Un test unitaire dédié pour le rollback serait redondant avec le test d'intégration du store.

File: frontend/projects/webapp/src/app/ui/dialogs/confirmation-dialog.ts [hunk 11]
```diff
@@ -5,78 +5,50 @@ import {
...
-            [attr.color]="data.confirmColor || 'primary'"
-            (click)="onConfirm()"
-            data-testid="confirm-delete-button"
-          >
-            {{ data.confirmText || 'Confirmer' }}
-          </button>
-        }
-      </mat-dialog-actions>
-    </div>
+    <mat-dialog-actions align="end" class="gap-2">
```
[the remark this answers is not in the document]
Comment (neogenz) [reply frontend/projects/webapp/src/app/ui/dialogs/confirmation-dialog.ts:28]: Merci 👍

File: shared/src/calculators/budget-formulas.ts [hunk 12]
```diff
@@ -167,35 +167,59 @@ export class BudgetFormulas {
...
+        const consumed = transactions
+          .filter(
+            (tx) =>
+              tx.budgetLineId === line.id &&
+              tx.checkedAt != null &&
+              (tx.kind === 'expense' || tx.kind === 'saving'),
+          )
+          .reduce((sum, tx) => sum + tx.amount, 0);
+
+        const effectiveAmount = Math.max(line.amount, consumed);
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts]: Les deux méthodes (`calculateTotalExpensesWithEnvelopes` et `calculateRealizedExpenses`) ont des conditions de filtrage différentes (l'une filtre sur `checkedAt`, l'autre non). L'extraction d'un helper commun ajouterait un paramètre `mustBeChecked` qui est un flag parameter — exactement ce que nos coding rules interdisent. Le code actuel est plus lisible tel quel.

File: shared/src/calculators/budget-formulas.spec.ts [hunk 13]
```diff
@@ -310,64 +342,377 @@ describe('BudgetFormulas', () => {
     });
   });
 
+  describe('calculateRealizedExpenses with envelope logic', () => {
+    function createBudgetLineWithId(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:373]: Merci !

File: shared/src/calculators/budget-formulas.spec.ts [hunk 14]
```diff
@@ -310,64 +342,377 @@ describe('BudgetFormulas', () => {
...
+      expect(result).toBe(200);
+    });
+
+    it('should not double-count with multiple transactions under envelope', () => {
+      // This is the main bug scenario from the issue
+      const budgetLines = [
+        createBudgetLineWithId('envelope-1', 'expense', 500, '2025-01-15'),
+      ];
+      const transactions = [
+        createTransactionWithBudgetLineId(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:590]: Merci !

File: shared/src/calculators/budget-formulas.ts [hunk 15]
```diff
@@ -207,8 +231,8 @@ export class BudgetFormulas {
    * @returns Solde calculé depuis les éléments cochés uniquement
    */
   static calculateRealizedBalance(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:230]: `calculateRealizedIncome` n'a pas besoin de `id` ni de `budgetLineId` car elle ne fait pas de logique d'enveloppe — elle somme simplement les montants des lignes income checkées. Le typage plus permissif est intentionnel ici : on n'a pas besoin de l'id pour ce calcul. Pas de risque de perte de garantie.

File: shared/src/calculators/budget-formulas.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:44]: `isRollover` est une propriété ajoutée dynamiquement par le frontend lors de la construction de la liste affichée (les lignes rollover sont des lignes virtuelles, pas des entités DB). L'ajouter au schéma Zod de `BudgetLine` serait incorrect car ce n'est pas une propriété qui existe en base. L'extension locale via interface est le bon pattern ici.

File: shared/src/calculators/budget-formulas.ts [hunk 16]
```diff
@@ -167,35 +168,55 @@ export class BudgetFormulas {
...
-      .reduce((sum, line) => sum + line.amount, 0);
+    let total = 0;
 
-    const checkedTransactionExpenses = transactions
-      .filter(
-        (t) =>
-          t.checkedAt != null && (t.kind === 'expense' || t.kind === 'saving'),
-      )
-      .reduce((sum, t) => sum + t.amount, 0);
+    budgetLines.forEach((line) => {
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:188]: La logique ternaire actuelle est concise et les tests couvrent exhaustivement tous les cas. L'ajout de variables intermédiaires augmenterait le nombre de lignes sans gain de clarté significatif — les commentaires inline suffisent.

File: shared/src/calculators/budget-formulas.spec.ts [hunk 17]
```diff
@@ -310,64 +369,428 @@ describe('BudgetFormulas', () => {
     });
   });
 
+  describe('calculateRealizedExpenses with envelope logic', () => {
+    function createBudgetLineWithId(
```
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:373]: Les helpers sont dans le même fichier de test et sont spécifiques aux types de ce module. Les extraire dans un fichier partagé ajouterait un couplage entre fichiers de test pour un gain minime. YAGNI.

File: shared/src/calculators/budget-formulas.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts]: On décline cette suggestion. La méthode proposée utilise un paramètre `mustBeChecked: boolean` qui est un flag parameter — ce qui viole directement notre convention clean-code (`save(isDraft: boolean)` → `saveAsDraft()` / `saveAndPublish()`). Les deux méthodes ont des critères de filtrage suffisamment différents (total vs realized) pour justifier l'implémentation séparée. Le couplage introduit par l'extraction rendrait les deux méthodes plus rigides sans gain de maintenabilité.

File: shared/src/calculators/budget-formulas.spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:373]: Merci 👍

File: shared/src/calculators/budget-formulas.spec.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:590]: Merci, c'était l'objectif — documenter le bug via le test.

File: shared/src/calculators/budget-formulas.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:230]: C'est intentionnel. `calculateRealizedIncome` n'a pas besoin de `id` ni `budgetLineId` — elle filtre uniquement par `kind` et `checkedAt`. Accepter `FinancialItem[]` (le supertype) rend la méthode plus réutilisable. Le passage de `FinancialItemWithId[]` vers `FinancialItem[]` est un upcast standard TypeScript (Liskov Substitution) — aucune donnée n'est perdue, et TypeScript le gère en toute sécurité.

File: shared/src/calculators/budget-formulas.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:44]: `isRollover` est une propriété d'enrichissement ajoutée dynamiquement côté frontend (dans `createRolloverLine()`). Elle n'a pas vocation à exister dans le schéma Zod de `BudgetLine` car les lignes rollover sont des entités virtuelles — elles n'existent pas en DB. L'interface `FinancialItemWithId` est locale au fichier de formules, ce qui est le bon scope pour cette extension.

File: shared/src/calculators/budget-formulas.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.ts:188]: La logique actuelle est déjà lisible en 4 lignes. Ajouter `isChecked`, `shouldUseEnvelope` et restructurer le if/else ajoute du code sans clarifier le comportement. De plus, le cas `isChecked && consumed === 0` est déjà correctement couvert par `Math.max(line.amount, consumed)` qui retourne `line.amount` quand `consumed === 0`.

File: shared/src/calculators/budget-formulas.spec.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply shared/src/calculators/budget-formulas.spec.ts:373]: Les helpers sont dans des `describe` blocks différents avec des signatures légèrement différentes (l'un inclut `checkedAt`, l'autre non). Les extraire dans un fichier séparé pour un seul fichier spec ajoute de l'indirection sans réduire significativement la duplication. On garde tel quel.

File: frontend/projects/webapp/src/app/ui/dialogs/confirmation-dialog.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (neogenz) [reply frontend/projects/webapp/src/app/ui/dialogs/confirmation-dialog.ts:28]: Merci 👍
